### PR TITLE
fixes #17173

### LINKED ISSUE #17173: strbasics.strip SIGBUS with --gc:arc
### Example
```nim
import std/strbasics
var a = "  vhellov   "
strip(a)
```

### Current Output
Traceback (most recent call last)
/Users/timothee/git_clone/nim/timn/tests/nim/all/t11905.nim(7) t11905
/Users/timothee/git_clone/nim/Nim_devel/lib/std/strbasics.nim(97) strip
/Users/timothee/git_clone/nim/Nim_devel/lib/std/strbasics.nim(67) setSlice
SIGBUS: Illegal storage access. (Attempt to read from nil?)

### Expected Output
works

### Additional Information
1.5.1 e9a287fe120501de4c4f70fbf687e0ceb9b75ee9

### PR DESCRIPTION
**Pull request by Araq**
fixes #17173

Todo: 

- [x] Documentation for `prepareStrMutation`
- [x] Add the test case.

Follow up

- [x] Documentation of ORC's COW strings

### CODE REVIEW

File: changelog.md [hunk 1]
```diff
@@ -173,6 +173,11 @@ provided by the operating system.
   dumping (on select signals) and notifying the parent process about the cause
   of termination.
 
+- Added `strip` and `setSlice` to `std/strbasics`.
```
Comment (ringabout) [inline changelog.md]:
```suggestion
```

File: lib/system.nim [hunk 2]
```diff
@@ -3126,3 +3126,6 @@ export io
 
 when not defined(createNimHcr) and not defined(nimscript):
   include nimhcr
+
+when not defined(gcDestructors):
```
Comment (ringabout) [inline lib/system.nim]:
```suggestion
when notJSnotNims and defined(nimSeqsV2):
```

File: lib/system/strs_v2.nim [hunk 3]
```diff
@@ -168,3 +168,8 @@ proc nimPrepareStrMutationImpl(s: var NimStringV2) =
 proc nimPrepareStrMutationV2(s: var NimStringV2) {.compilerRtl, inline.} =
   if s.p != nil and (s.p.cap and strlitFlag) == strlitFlag:
     nimPrepareStrMutationImpl(s)
+
+proc prepareStrMutation*(s: var string) {.inline.} =
```
Comment (ringabout) [inline lib/system/strs_v2.nim:172]:
```suggestion
proc prepareStrMutation*(s: var string) {.inline.} =
  # string literals are "copy on write", so you need to call
  # `prepareStrMutation` before modifying the strings.
```

File: lib/system.nim [hunk 4]
```diff
@@ -3126,3 +3126,6 @@ export io
...
+  proc prepareStrMutation*(s: var string) {.inline.} = discard
```
Comment (ringabout) [inline lib/system.nim]:
```suggestion
  proc prepareStrMutation*(s: var string) {.inline.} =
    ## String literals(i.e. "abc", etc) in ARC/ORC mode are "copy on write",
    ## therefore you should call `prepareStrMutation` before modifying the strings.
    discard
```

Comment (Araq) [reply lib/system.nim]: "before modifying the strings via `addr`"

File: lib/std/strbasics.nim [hunk 5]
```diff
@@ -64,11 +64,13 @@ func setSlice*(s: var string, slice: Slice[int]) =
       when not declared(moveMem):
         impl()
       else:
+        when defined(gcDestructors):
```
Comment (ringabout) [inline lib/std/strbasics.nim]:
```suggestion
        when defined(nimSeqsV2):
```

File: tests/arc/t17173.nim [hunk 6]
```diff
@@ -0,0 +1,10 @@
+discard """
+  cmd: "nim c -r --gc:arc $file"
```
Comment (ringabout) [inline tests/arc/t17173.nim]:
```suggestion
  matrix: "--gc:refc; --gc:arc; --newruntime"
```

File: lib/system.nim [hunk 7]
```diff
@@ -3126,3 +3126,9 @@ export io
 
 when not defined(createNimHcr) and not defined(nimscript):
   include nimhcr
+
+when notJSnotNims and defined(nimSeqsV2):
```
Comment (ringabout) [inline lib/system.nim]:
```suggestion
when notJSnotNims and not defined(nimSeqsV2):
```

File: lib/system.nim [hunk 8]
```diff
@@ -3126,3 +3126,9 @@ export io
 
 when not defined(createNimHcr) and not defined(nimscript):
   include nimhcr
+
+when notJSnotNims and not defined(nimSeqsV2):
+  proc prepareStrMutation*(s: var string) {.inline.} =
+    ## String literals(i.e. "abc", etc) in ARC/ORC mode are "copy on write",
```
Comment (ringabout) [inline lib/system.nim]:
```suggestion
    ## String literals(e.g. "abc", etc) in the ARC/ORC mode are "copy on write",
```

Comment (ghost) [reply lib/system.nim]: While this change is fine (most people will understand what it means anyway), `i.e.` is correct here as well - https://en.wiktionary.org/wiki/i.e. means "that is", so the sentence reads like "String literals (that is - "abc", etc)".

File: lib/system.nim [hunk 9]
```diff
@@ -3126,3 +3126,9 @@ export io
 
 when not defined(createNimHcr) and not defined(nimscript):
   include nimhcr
+
+when notJSnotNims and not defined(nimSeqsV2):
+  proc prepareStrMutation*(s: var string) {.inline.} =
+    ## String literals(e.g. "abc", etc) in the ARC/ORC mode are "copy on write",
```
Comment (ghost) [inline lib/system.nim]:
```suggestion
    ## String literals (e.g. "abc", etc) in the ARC/ORC mode are "copy on write",
```
Just a nitpick, space before the paren :)

Comment (ringabout) [reply lib/system.nim]: thanks

**Comment by Araq** [comment]
Unrelated CI failure, merging.

File: lib/system.nim (hunk 4 above)
Comment (timotheecour) [reply lib/system.nim]: this definitely needs a `runnableExamples`

Comment (ringabout) [reply lib/system.nim]: ok

Comment (timotheecour) [reply lib/system.nim]: => https://github.com/nim-lang/Nim/pull/17227

**Pull request merged**